### ISSUE TITLE
Support checking against WCAG 2.1

### ISSUE THREAD

**Issue by avfall98** [squizlabs/HTML_CodeSniffer#214]
I believe this spec is not finalised and out yet, but raising an issue now for it so that we can start to investigate what the effort is and eventually schedule in development to get it supported.

https://www.w3.org/TR/WCAG21/

**Comment by ironikart** [comment]
This is on our radar and we'll definitely be adding support for those new criteria where compliance is able to be tested via script. As an observation, there are quite a few criteria in the current 2.1 candidate spec that will probably place more emphasis on human testing.

**Comment by katannshaw** [comment]
@ironikart: Thank you for working on this issue. Is there an update?

**Comment by ironikart** [comment]
Hi @katannshaw. We are still working on this but hope to have it in the next release.

**Comment by jameelmoses** [comment]
Hi @ironikart, any updates on this?

**Comment by ironikart** [comment]
Hi @jameelmoses. We've scoped out the work but I haven't had time myself to do this amongst other projects. I'll probably need to get someone else to help with this feature.

**Comment by reidelliott** [comment]
@ironikart, how can we help?

**Comment by rakeshkmr516** [comment]
@ironikart: Could you please work on the same or assign someone

**Comment by ironikart** [comment]
Hi @reidelliott, @rakeshkmr516, apologies for the late reply. In the scope of work we identified that many of the new criteria for WCAG 2.1 had elements we couldn't detect programatically. Some criteria have methods for achieving partial detection programatically and would need to be broken up to report separately. In HTMLCS we display a warning when a check requires manual intervention. 

What we need to do is write sniffs for the following:

1.3.6 - Identify Purpose
- Check for No <main> or role, aria-role main

1.4.10 - Reflow
- Detect element font-size using px or pt and display a warning
- Display an error for user-scalable=no or user-scalable=0 in the <meta> tags
- Investigate trapping of scroll in <pre>, if it can be programatically checked display a warning

1.4.12 - Text Spacing
- (complex) - Write a sniff for font-size, line-height and letter-spacing and report looking for overlapping letters or words or text otherwise made unreadable. Subjective, particularly with different language character sets, a warning may be sufficient.

2.5.3 - Label in Name
- Detect if the accessible label of <label> matches the visible text content
- Detect if the accessible label of <label> matches the alternate text content of an image
- Display a warning for manual checking for labels affecting visible image content

---

The rest of the research we've done resulted in either in the tests being very subjective (e.g. document complexity), or not being able to accurately detect an issue. The rest of the criteria would have to be formed into warnings that call out for manual checks to be performed.

1.3.6 - Identify Purpose
- Relative document complexity (subjective)
- Check if aria-invalid or aria-required needed by inputs
- Whether the image used in <img> or background is a symbol that should be marked up to allow the user to substitute replacements
- Whether the role or aria attribute has been applied correctly

1.4.10 - Reflow
- Browser zoom (unable to detect) - Generic warning to check how the document appears when zoom is used

1.4.11 - Non Text Contrast
- Warnings for graphic elements that contrast cannot be determined and must be manually checked (transparency, positioning - this has come up before with other contrast tests)

1.4.13 - Content on hover or focus
2.4.1 - Character key shortcuts
2.2.6 - Timeouts
2.3.3 - Animations from Interactions
2.5.1 - Pointer Gestures
2.5.2 - Pointer Cancellation
2.5.4 - Motion Actuation
2.5.5 - Target Size (Can’t accurately detect if the element is a target for pointer input, i.e. event listeners, different pointer devices).
2.5.6 - Concurrent input mechanisms
4.1.3 - Status Messages (Can’t detect if content is a status message and requires a role or which role to contextually apply).

**Comment by downcast** [comment]
@ironikart, thanks for scoping out the work. Do you have an eta on how long you think it would take before we see it in a release? As much as I love forking and coding updates myself, this seems best left to pros.

**Comment by ironikart** [comment]
@downcast Sorry no eta. Depending how far we take support for it some of the work represents a significant chunk of time. At least a week, but my gut feeling probably more like 2-3 if we tackle everything we can test with a decent level of fidelity. Right now I'm only able to break from my main work priorities to fix bugs or review pull requests with the project.

**Comment by reidelliott** [comment]
@ironikart I think you're on the money when you said
> ... As an observation, there are quite a few criteria in the current 2.1 candidate spec that will probably place more emphasis on human testing.

Very much so! There's a lot to WCAG 2.1 that requires some common sense and human testing. While some parameters may pass compliance, they still might not be accessible in the sense of usable or not.

**Comment by bdeclerc** [comment]
We've implemented a number of WCAG 2.1 checks in our fork - https://github.com/openfed/AccessibilityCheck (you can check an overview of coverage on https://openfed.github.io/AccessibilityCheck/overview/index.html )

I'm looking for budget for the next phase in enhancement of our tool (including re-syncing bugfixes & changes with HTML Codesniffer) but in the mean time, feel free to take stuff from our fork.

**Comment by stefanruijsenaars** [comment]
PR: https://github.com/squizlabs/HTML_CodeSniffer/pull/271

**Comment by stefanruijsenaars** [comment]
Now that this has been merged, feel free to close :)
